File: 백준/Bronze/30614. Port Robot/Port Robot.cc
#define BOJ_SUBMIT
#ifndef BOJ_30000_SRC_30614_PORT_ROBOT_H_
#define BOJ_30000_SRC_30614_PORT_ROBOT_H_

#include <iostream>
#include <stack>
#include <string>

class PortRobot {
 public:
  void SetInputs() {
    std::cin >> log_length;
    std::cin >> logs;
  }

  void Calculate() {
    result = false;
    if (log_length % 2) {
      return;
    }

    const char diff = 'a' - 'A';

    for (size_t i = 0; i < log_length; ++i) {
      if ('a' <= logs[i] && logs[i] <= 'z') {
        log_stack.push(logs[i]);
      } else {
        if (log_stack.empty()) {
          return;
        }
        if ((log_stack.top() - diff) == logs[i]) {
          log_stack.pop();
        } else {
          return;
        }
      }
    }
    result = log_stack.empty();
  }

  void PrintResult() { std::cout << result; }

 private:
  size_t log_length;
  std::string logs;
  std::stack<char> log_stack;
  bool result;
};

#ifdef BOJ_SUBMIT
int main() {
  PortRobot port_robot;
  port_robot.SetInputs();
  port_robot.Calculate();
  port_robot.PrintResult();

  return 0;
}
#endif  // BOJ_SUBMIT

#endif  // BOJ_30000_SRC_30614_PORT_ROBOT_H_

File: 백준/Bronze/30614. Port Robot/Port Robot.py
log_length = int(input())
logs = input().strip()

def calculate():
    log_stack = []
    result = False
    if (log_length % 2):
        return result
    
    diff = ord('a') - ord('A')
    
    for log in logs:
        if 'a' <= log <= 'z':
            log_stack.append(log)
        else:
            if not log_stack:
                return result
            
            if (ord(log_stack[-1]) - diff) == ord(log):
                log_stack.pop()
            else:
                return result
    result = not log_stack
    return result

if __name__ == "__main__":
    if calculate():
        print(1)
    else:
        print(0)
